azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java

```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
       setTestOnBorrow(true);
     }
 
+    /**
+     * This method overrides {@link BasicDataSource#getConnection()}, in order to have retry logics.
+     *
+     */
+    @Override
+    public Connection getConnection() throws SQLException {
+
+      /*
+       * when az server process launches, if the connection isn't valid (e.g., network problem, wrong password, etc),
+       * we fail the connection immediately.
+       */
+      if (getDataSource() == null) {
+        return createDataSource().getConnection();
+      }
+
+      Connection connection = null;
+      int retryAttempt = 0;
+      this.dataSource = null;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
```

Comment:
This method works all ok - but usually its a good idea to try exponential/ fibonacci backoff in such cases

This is a good read: https://carlosbecker.com/posts/exponential-backoff-java8/
This as well: https://github.com/rholder/guava-retrying

We can keep it simple and not necessarily have to use external dependency - but this is a good example of using exponential backoff